app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailFragment.kt

```diff
@@ -22,86 +22,71 @@ import android.view.MenuInflater
-        setupFab()
-        view.setupSnackbar(this, viewModel.snackbarText, Snackbar.LENGTH_SHORT)
-        setupNavigation()
-        this.setupRefreshLayout(viewDataBinding.refreshLayout)
-    }
-
-    private fun setupNavigation() {
-        viewModel.deleteTaskEvent.observe(
-            this,
-            EventObserver {
-                val action = TaskDetailFragmentDirections
-                    .actionTaskDetailFragmentToTasksFragment(DELETE_RESULT_OK)
-                findNavController().navigate(action)
-            }
-        )
-        viewModel.editTaskEvent.observe(
-            this,
-            EventObserver {
-                val action = TaskDetailFragmentDirections
-                    .actionTaskDetailFragmentToAddEditTaskFragment(
-                        args.taskId,
-                        resources.getString(R.string.edit_task)
-                    )
-                findNavController().navigate(action)
-            }
-        )
-    }
-
-    private fun setupFab() {
-        activity?.findViewById<View>(R.id.edit_task_fab)?.setOnClickListener {
-            viewModel.editTask()
-        }
-    }
-
     override fun onCreateView(
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?
-    ): View? {
-        val view = inflater.inflate(R.layout.taskdetail_frag, container, false)
-        viewDataBinding = TaskdetailFragBinding.bind(view).apply {
-            viewmodel = viewModel
-        }
-        viewDataBinding.lifecycleOwner = this.viewLifecycleOwner
-
-        viewModel.start(args.taskId)
-
+    ): View {
         setHasOptionsMenu(true)
-        return view
+
+        return ComposeView(requireContext()).apply {
+            // Dispose of the Composition when the view's LifecycleOwner is destroyed
+            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+            setContent {
+                AppCompatTheme {
+                    Surface(Modifier.fillMaxSize()) {
+                        TaskDetailScreen(
+                            taskId = args.taskId,
+                            viewModel = viewModel,
+                            onEditTask = {
+                                val action = TaskDetailFragmentDirections
+                                    .actionTaskDetailFragmentToAddEditTaskFragment(
+                                        args.taskId,
+                                        resources.getString(R.string.edit_task)
+                                    )
+                                findNavController().navigate(action)
+                            }
+                        )
+                    }
+                }
+            }
+        }
     }
 
     override fun onOptionsItemSelected(item: MenuItem): Boolean {
         return when (item.itemId) {
             R.id.menu_delete -> {
-                viewModel.deleteTask()
+                lifecycleScope.launch {
```

Comment:
This new scoping here feels a bit weird to me but maybe it's ok because I am not a coroutine expert. Wouldn't this now cause a delay while it waits to delete the task? I'm sure it's super quick but given this is an architecture sample would it be better to architect this as if it was slow like a network call? 

The other thing that feels wrong is that navigate call would have the effect of cancelling the scope of this coroutine, is that fine?

Reply:
Yeah, that's fine. By the time the navigation call happens, the task is already deleted since `deleteTask` is a suspend function.

> Wouldn't this now cause a delay while it waits to delete the task?
Agreed, but I'm not modifying ViewModel business logic in this migration, but if the `deleteTask` call could take a long time, the ViewModel should update the `uiState` to indicate that something is in progress. Maybe that's something to fix in the future as the same thing could've happened before.

Reply:
All good!